### DISCUSSION TITLE
JAVA_HOME does not point to a valid JDK installation

### DISCUSSION THREAD

**Discussion by bobsmo** [airsdk/Adobe-Runtime-Support#1695] [Q&A]
Compiling on Mac 
I get this error.

JAVA_HOME does not point to a valid JDK installation, which is required for Android App Bundle creation

how do i find where that home is on my. mac ?
so i can edit. the adt.cfg ?

# JAVA_HOME is usually picked up from an environment variable but this can override it:
#JAVA_HOME= ???

**Comment by ajwfrost** [comment]
If you open a terminal window you can do a full-machine search for "javac", the Java compiler that is in the JDK: see below for something like that here:

```
% find / -name "javac" 2>/dev/null
/usr/bin/javac
/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/bin/javac
/System/Volumes/Update/mnt1/usr/bin/javac
/System/Volumes/Data/Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/bin/javac
/System/Volumes/Data/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/javac
/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/javac
```

So you can ignore the `/System` ones, and the `/usr/bin/` ones aren't so useful: I could then use with the OpenJDK 1.8 or the Android Studio-provided versions... I would normally go with Android Studio so the 'home' part of this would be:
```
JAVA_HOME=/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home
```

thanks

**Reply by bobsmo** [reply]
thanks for the help. 


 I tried both  

/Library/Java/JavaVirtualMachines/jdk1.8.0_102.jdk/Contents/Home/bin/javac
and
/Applications/Android Studio.app/Contents/jre/Contents/Home/bin/javac

also, I added the 

```
<android>
        <BuildLegacyAPK>true</BuildLegacyAPK>
        <manifestAdditions><![CDATA[<manifest><uses-permission android:name="android.permission.INTERNET"/></manifest>]]></manifestAdditions>
</android>
```

 and got these errors.

WARNING: could not find appropriate Java Home location for v11+ JDK
unexpected failure: Unable to run aapt
java.io.IOException: Unable to run aapt
	at com.adobe.air.apk.APKOutputStream.a(APKOutputStream.java)
	at com.adobe.air.apk.APKOutputStream.addApplicationDescriptor(APKOutputStream.java)
	at com.adobe.air.ApplicationPackager.addSpecialFiles(ApplicationPackager.java)
	at com.adobe.air.ApplicationPackager.createPackage(ApplicationPackager.java)
	at com.adobe.air.ADT.parseArgsAndGo(ADT.java)
	at com.adobe.air.ADT.run(ADT.java)
	at com.adobe.air.ADT.main(ADT.java)

**Reply by ajwfrost** [reply]
Not sure whether the first of the paths would be available on your machine? I was hoping that you could run the "find" command to see if/where you have a JDK installation (and then you'd need the JAVA_HOME variable to point to the "Home" folder rather than to the Home/bin/javac executable).

But if we focus on getting the legacy build to work first: the failure there may be an issue with it using a newer AAPT version. I can't see why the JDK 11 warning would come out if it's correctly picked up the BuildLegacyAPK flag, and yet it's definitely using the legacy mechanism (APKOutputStream). But it's failing when running AAPT (which is an executable file, not needing Java)...

Are you able to comment out the `AndroidPlatformSDK` value from the config file? (or check whether Animate is passing one?) as there may be an issue with this: it would be good to know what Android SDK tools are installed. If you can also uncomment `DebugOut=true` then we should be able to see a log file (adt.log) generated in your home folder, which should help provide more information..

thanks

**Reply by bobsmo** [reply]
Thanks a million . 
I commented out the #AndroidPlatformSDK=   
That seemed to do the trick.  Not sure what the hell is going on... but I'll take it.

I also left in the  `<BuildLegacyAPK>true</BuildLegacyAPK>`
and also   JAVA_HOME=/Applications/Android Studio.app/Contents/jre/Contents/Home

Currently just building and installing on test platforms... So not sure if any of this will apply when I try to upload to google play store.

**Reply by ajwfrost** [reply]
To upload to Play Store you're going to need an Android App Bundle, and for that it will be back to square one with the need for setting up platform (Android) SDK and Java home folder. We had made some updates in the latest builds of AIR SDK which attempt to do more to find these paths out for you, based on default installation locations of Android Studio and the Android SDK, but there are still instabilities in those SDKs hence they're not the "latest release" version that you get from the download page.. hopefully within a week or two, we'll have resolved the last problem, although another way to do it would be to take the "adt.jar" file from the latest SDK (33.1.1.779) and copy this over the top of the one from 33.1.1.743 or whatever you're using..

thanks

**Reply by SporeProductions** [reply]
Getting this error with AIR SDK 50.2.3.4 
Has this not been fixed?

UPDATE: Need to use more recent version of Animate

**Reply by ajwfrost** [reply]
Not sure there was anything to fix for this one. If you're getting an error message about JAVA_HOME not pointing to a valid JDK installation, can I check which message you get? If it's about App Bundle creation, it means that we're not able to find a JDK of version 11 or higher, using the configuration settings. If it's about App Bundle signing, we can't find a JDK of version 8 or higher.

The message actually comes out if we can't find the appropriate file for these activities:
`JAVA_HOME/bin/javac` for creating the bundle
`JAVA_HOME/bin/jarsigner` for signing it

The way we look for 'JAVA_HOME' is, in decreasing priority:
1) if you have this setting defined in the adt.cfg file in your `~/.airsdk` folder
2) if you have this setting defined in the adt.cfg file in your SDK `lib` folder
3) checking in the currently-running Java virtual machine location
4) checking for the `JAVA_HOME` environment variable
5) trying to see if we can find an installation of Android Studio

Hopefully that will help you work out why the error message is coming out? Looking at it, I'm not sure we have specific debug messages that would definitely identify the location that's being checked for the javac or jarsigner applications.. but we can add those in to the output log for the future, to help solve this issue quicker.

thanks

**Reply by SporeProductions** [reply]
Was using an old version of Adobe Animate (2019) and the errors went away when publishing/packaging with Animate 2020
